uc-crux-llvm/test/Test.hs

```diff
@@ -110,19 +112,47 @@ _exampleStruct =
 testDir :: FilePath
 testDir = "test/programs"
 
+data UCCruxLLVMTestLogMessage
+  = ClangTrouble
+
+sayUCCruxLLVMTest ::
+  Crux.Logs msgs =>
+  ( ?injectUCCruxLLVMTestLogMessage :: UCCruxLLVMTestLogMessage -> msgs ) =>
+  UCCruxLLVMTestLogMessage ->
+  IO ()
+sayUCCruxLLVMTest msg =
+  let ?injectMessage = ?injectUCCruxLLVMTestLogMessage
+   in Crux.say msg
+
+ucCruxLLVMTestLogMessageToSayWhat ::
+  UCCruxLLVMTestLogMessage ->
+  Crux.SayWhat
+ucCruxLLVMTestLogMessageToSayWhat ClangTrouble =
+  Crux.SayWhat Crux.Fail "UC-Crux-LLVM" "Trouble when running Clang:"
+
 findBugs ::
   Maybe L.Module ->
   FilePath ->
   [String] ->
   IO (Map.Map String Result.SomeBugfindingResult)
 findBugs llvmModule file fns =
   do
+    -- log structure is "Crux + (CruxLLVM + UCCruxLLVMTest)"
+    let ?injectCruxLogMessage = Left
+    let ?injectCruxLLVMLogMessage = Right . Left
+    let ?injectUCCruxLLVMTestLogMessage = Right . Right
+    let logMessageToSayWhat =
+          either Crux.cruxLogMessageToSayWhat
+            (either
+              Log.cruxLLVMLogMessageToSayWhat
+              ucCruxLLVMTestLogMessageToSayWhat)
```

Comment:
I'm still concerned that this performs the conversion at the caller side, rather than the callee side, which will mean that if we use a different `LogAction` for the IDE that it will only be receiving `SayWhat` messages and have lost a lot of the structural info that would be useful.

My suggestion would be to create the following here:

    data UCCruxLLVMTestLogMsg = UCCruxLLVMLog UCCruxLLVMTestLogMessage
                              | CruxLLVMLog CruxLLVMLogMessage

Then parameterize `OutputConfig` by `msg` (where here in uc-crux-llvm Test.hs, `msg` is `UCCruxLLVMLog`) and require the initialization point (like here in Test.hs) to provide a `_logFrontEnd :: LogAction IO msg` (where your default can be a contravariant functor that converts to `SayWhat` that is applied to the `SayWhat` log action, but the initialization could change that to be an IDE-related thing.  

In Crux.Log you can then have:


    class  ToLogMsg o m where
        toLogMsg :: o -> m

    sayMsg :: (ToLogMsg o msg, Logs msg) => o -> IO ()
    sayMsg = LJ.writeLog (_logFrontEnd ?outputConfig) . toLogMsg

Using this approach, the caller uses `sayMsg` which defers any conversion to the `_logFrontEnd` log action, and that allows the initialization code to setup the type of log action that is appropriate for its configuration (default to user, JSON to IDE, etc.).  

Each front end then provides a `ToLogMsg` instance that helps convert other Crux module messages into the `msg` type.  For example, this uc-crux-llvm Test.hs might have:

    instance ToLogMsg UCCruxLLVMTestLogMessage UCCruxLLVMTestLogMsg
        where toLogMsg = UCCruxLLVMLog

     instance ToLogMsg CruxLLVMLogMessage UCCruxLLVMTestLogMsg
        where toLogMsg = CruxLLVMLog

which allows crux-llvm to log it's own messages as long as there is a ToLogMsg class constraint for its messages.

I'm sure that there's a hole or two in this approach, but the main advantage is allowing the initialization point to determine how the logged messages are converted and emitted (or not) rather than having to spread that information to all the call sites and only providing the actual log emitter with a generic form.